src/ioctls/vcpu.rs

```diff
@@ -1569,7 +1578,33 @@ impl VcpuFd {
                 r => Ok(VcpuExit::Unsupported(r)),
             }
         } else {
-            Err(errno::Error::last())
+            let errno = errno::Error::last();
+            // From https://docs.kernel.org/virt/kvm/api.html#kvm-run :
+            //
+            // KVM_EXIT_MEMORY_FAULT is unique among all KVM exit reasons in that it accompanies
+            // a return code of ‘-1’, not ‘0’! errno will always be set to EFAULT or EHWPOISON
+            // when KVM exits with KVM_EXIT_MEMORY_FAULT, userspace should assume kvm_run.exit_reason
+            // is stale/undefined for all other error numbers.
+            if ret == -1
+                && (errno == errno::Error::new(libc::EFAULT)
+                    || errno == errno::Error::new(libc::EHWPOISON))
+            {
+                let run = self.kvm_run_ptr.as_mut_ref();
+                if run.exit_reason == KVM_EXIT_MEMORY_FAULT {
```

Comment:
Maybe combine these 2 `if`?
```Rust
if ret == -1
  && (errno == errno::Error::new(libc::EFAULT) || errno == errno::Error::new(libc::EHWPOISON)) 
  && self.kvm_run_ptr.as_mut_ref().exit_reason == KVM_EXIT_MEMORY_FAULT
            {
                ...
```
This way we will not have 2 `else`s with `Err(errno)`

Reply:
in that case we'd have to dereference `self.kvm_run_ptr` twice: once for the condition and once for fault data. Do you think it's better than 2 elses?